cmdline arg to convert DWI to SSC+SM (simple wrapper around song->load and song->save)

teleshoes opened the pull request:
quick and dirty cmdline wrapper. feel free to close it, just PR'ing in case someone else finds it useful.
converted 280 old DWI files to ssc in about 5s, seems to work well

`Usage: ./stepmania --convert Songs/MyPack/MySong`


wrapper i used to convert all my dwi-only song dirs:
https://github.com/teleshoes/wolke-home-config/blob/e40f437db7649a026891df0266d904d5a0645c1a/bin/stepmania-ensure-ssc-sm

teleshoes committed:
add cmdline opt --convert SONG_DIR to read sm/ssc/dwi/etc and write to ssc+sm

Horsey- commented:
I would be VERY interested in a similar tool to batch generate SSC files next to SM counterparts.

teleshoes commented:
this tool already does that. just run `./stepmania --convert Songs/MyGroup/MySong` for a song where the *.sm exists and the *.ssc does not
dont run it if the ssc already exists tho

teleshoes commented:
this command line tool is almost exactly the same as opening the song up on stepmania editor and hitting save without making any changes

sillybear commented:
I think it's better to convert dwi to sm only since dwi uses song timing. 

ssc files are best used for those that used steps timing.

Horsey- commented:
> this tool already does that. just run `./stepmania --convert Songs/MyGroup/MySong` for a song where the *.sm exists and the *.ssc does not
> dont run it if the ssc already exists tho

Will it work if I point it to a pack rather than just a single song folder? I’d like to basically convert whole packs at once.

teleshoes commented:
you see my perl script in the comment? you can wrap this tool easily enough in something like it. this is less efficient than looping in stepmania, but it can handle hundreds of dirs per minute for me. my goal was minimal change and least brittle, not most feature-rich or most performant

Horsey- commented:
> you see my perl script in the comment? you can wrap this tool easily enough in something like it. this is less efficient than looping in stepmania, but it can handle hundreds of dirs per minute for me. my goal was minimal change and least brittle, not most feature-rich or most performant

I’m very sorry to be a PITA, but I have zero knowledge of perl; can you give an example on how to run that to generate SSC files?

teleshoes commented:
sure. save the file to `/tmp/a.pl`
change this line:
`my $CONVERTER_BUILD = "$ENV{HOME}/Games/stepmania/converter";`
=>
`my $CONVERTER_BUILD = "/opt/stepmania";`
or wherever you have the stepmania executable installed

make sure that `Songs` is your songs dir, and that its a sister of the executable

then just `perl /tmp/a.pl`

main point is that wrapping this tool in any arbitrary script is trivial to do, and thus theres no benefit to moving that logic into the c++ (except performance, which would be orders of magnitude better). heres a bash script:
`cd stepmania; for x in Songs/*/*; do ./stepmania --convert "$x"; done`

Horsey- commented:
so last night I built this on OS X and the build went through, but when I run the app, it loads but freezes 5 at this point. 
![Screen Shot 2020-11-10 at 10 41 49 AM](https://user-images.githubusercontent.com/5679966/98710805-60dee980-2341-11eb-8bde-1bd0ae810752.png)

teleshoes commented:
there is a 0% chance that this PR caused that, since it literally only affects commandline arg parsing.

try and build this PR's parent commit and see if you can reproduce.

Horsey- commented:
okay, I've been playing with this for the past week and got it to work in Ubuntu and OS11. I need to point out though that there was a typo in your OP

`Usage: ./stepmania --convert Songs/MyPack/MySong/` 
this terminal code makes stepmania panic and crash (the final forward slash shouldn't be there)

`Usage: ./stepmania --convert Songs/MyPack/MySong`
this terminal code works just fine 

I want to point out that this doesn't save over any .sm .ssc files as long as no .old files exist in the song folder. In which case, the .old files are overwritten and replaced (this is how it would work if you saved the song in the editor)

~~I'm going to continue playing with this to make it work in OSX. Would you happen to know how to invoke stepmania in the OSX terminal? "./stepmania" doesn't work in osx because it's probably not the correct syntax for that operating system.~~

For Mac OSX/OS11 (SM works perfectly fine in OS11) terminal: 
```
open StepMania.app --args --convert Songs/MySongs/MySong
```

I wanna just thank you for this btw.

Horsey- commented:
@teleshoes 

okay, so as above things are working, but the perl script doesn't work, I assume because you're running it from Linux? I tried changing things to the OS X terminal syntax, but it didn't work

```
#!/usr/bin/perl
use strict;
use warnings;

my $CONVERTER_BUILD = "/Users/zack/Stepmania/StepMania\ 5.1\ cmd\ line";

sub main(@){
  die "ERROR: $CONVERTER_BUILD not found\n" if not -d "$CONVERTER_BUILD/";
  chdir "$CONVERTER_BUILD/";
  $ENV{PWD} = "$CONVERTER_BUILD/";

  my @songDirs = glob "Songs/*/*/";
  for my $songDir(@songDirs){
    my @sm = (glob("$songDir/*.sm"), glob("$songDir/*.SM"));
    my @ssc = (glob("$songDir/*.ssc"), glob("$songDir/*.SSC"));
    my @dwi = (glob("$songDir/*.dwi"), glob("$songDir/*.DWI"));
    if(@dwi > 0 and @sm == 0 and @ssc == 0){
      print "$songDir\n";
      system "open StepMania.app", "--args --convert", $songDir;
    }
  }
}

&main(@ARGV);

```

teleshoes commented:
yay! im glad things are working correctly cross-platform. thanks for the fix you described (both versions work on linux, ill change it to the one that works on macosx)

about that perl script; sorry, its not intended to be portable, or actually to be used by anyone but me. its just an example usage for how you COULD script this if you like. if you write a script thats useful to you on macosx, please paste it here so that folks might find it.
on linux, i modified mine eventually `https://github.com/teleshoes/wolke-home-config/blob/master/bin/stepmania-ensure-ssc-sm`

as for a portable solution, i could build better cli syntax with multiple files, or a --all-songs option, or a --pack=FolderName arg to do something like i have, and way faster, but i dont have the time to compile&test it, now that the basic functionality works

teleshoes commented:
in case anyone is interested, here is a somewhat related stand-alone project (work-in-progress):
https://github.com/teleshoes/ddr-stepmania-tools/blob/master/src/simfile-radar

i actually added this cli flag so that i could process just SM files to calculate noteset statistics and radar values

```
> simfile-radar GAIA.sm
singles | beginner   | S22  | V22  | A12  | F8   | C0   |
singles | basic      | S55  | V52  | A15  | F37  | C12  |
singles | difficult  | S79  | V74  | A27  | F25  | C46  |
singles | expert     | S122 | V109 | A0   | F0   | C108 |
doubles | basic      | S55  | V52  | A24  | F37  | C12  |
doubles | difficult  | S76  | V59  | A26  | F21  | C40  |
doubles | expert     | S123 | V109 | A0   | F0   | C108 |

> simfile-radar --singles --expert --format=%07.2f-STREAM%n GAIA.sm
0122.32
```

Horsey- commented:
> as for a portable solution, i could build better cli syntax with multiple files, or a --all-songs option, or a --pack=FolderName arg to do something like i have, and way faster, but i dont have the time to compile&test it, now that the basic functionality works

If there's anything I can do to make it easier for you, let me know. I can compile and test for for whatever system you need.

sillybear commented:
@teleshoes After looking to your previous comment above, I'm curious on how you would implement/update the old groove radar calculation in SM's NoteDataUtil.cpp even if it is just an unofficial patch. Hopefully the DDR A-type calculation. We need to get out of the old groove radar calculation.

teleshoes commented:
sorry to say that i (personally) would not. the above script (unlike this somewhat related pr) is fully standalone, and just takes an `sm` or `ssc` file and parses it. the C++ code, especially NoteDataUtil.cpp, is a bit of a nightmare and would have taken me weeks to do, instead of hours.

by all means, feel free to use the simfile-parsing-algorithms (which are mine) and the numeric formulas (which come from https://dancedancerevolution.fandom.com/wiki/Groove_Radar) as a guide to rewrite the cpp. ive tested across hundreds of simfiles, and so far it reflects the DDR-A groove radar very closely (but not precisely; this is almost certainly because of actual differences in the simfile versus the official game).

...or are suggesting that you wanna bundle my script and a perl interpreter and literally make a system call to it? this is crazy, in my opinion, but i wouldnt necessarily be opposed

sillybear commented:
> in case anyone is interested, here is a somewhat related stand-alone project (work-in-progress):
> https://github.com/teleshoes/wolke-home-config/blob/master/bin/simfile-radar
> 
> i actually added this cli flag so that i could process just SM files to calculate noteset statistics and radar values
> 
> ```
> > simfile-radar GAIA.sm
> singles | beginner   | S22  | V22  | A12  | F8   | C0   |
> singles | basic      | S55  | V52  | A15  | F37  | C12  |
> singles | difficult  | S79  | V74  | A27  | F25  | C46  |
> singles | expert     | S122 | V109 | A0   | F0   | C108 |
> doubles | basic      | S55  | V52  | A24  | F37  | C12  |
> doubles | difficult  | S76  | V59  | A26  | F21  | C40  |
> doubles | expert     | S123 | V109 | A0   | F0   | C108 |
> 
> > simfile-radar --singles --expert --format=%07.2f-STREAM%n GAIA.sm
> 0122.32
> ```

How do you compile or run this for Windows?

teleshoes commented:
run it in putty, maybe? its just a perl script, but it assumes a unix-like set of tools are available, rm, mkdir, cat, etc, plus it uses $HOME/.cache/ as a dir. dunno what perl does with that on windows.

it should be trivial to port, tho. i dont have a windows machine, but id happily accept a pull request adding windows support if you get it working.

if you do `simfilar-radar --no-cache` it might even just work on windows, who knows

sillybear commented:
I tried to run the simfile-radar alone w/o the sm file and it displays all the info. Then I tried adding an sm file in the cmd line (with the actual file in the same location as the pl file, used GAIA.sm for this example), it says:

ERROR: sha256sum failed for GAIA.sm
No such file or directory

How to fix this?

teleshoes commented:
you need to pass it the actual SM file, so maybe something like:
`simfile-radar --no-cache Songs/Gaia/Gaia.sm`
`simfile-radar --no-cache "C:\Program Files/Stepmania\Songs\Gaia\GAIA.sm`
or whatever

EDIT: oh you said it already was in the same dir, sorry. so yea, just try `--no-cache` like i said, that should bypass (most of) the gnu/linux specific stuff

sillybear commented:
I just tried simfile-radar --no-cache GAIA.sm but it says

ERROR: unknown arg --no-cache

where can I find the latest copy of script? The one I got was from yours that got deleted(?)
![SharedScreenshot](https://user-images.githubusercontent.com/6451629/104530611-4bd74d00-5647-11eb-8ba8-08a5624de083.jpg)

teleshoes commented:
lol, sorry. i updated the link above. i forgot that i moved all this crap to its own repo
https://github.com/teleshoes/ddr-stepmania-tools/blob/master/src/simfile-radar

sillybear commented:
I updated my copy but still getting the same result.
![pic](https://user-images.githubusercontent.com/6451629/104538731-644f6380-5657-11eb-90d8-8be28cb28796.jpg)

teleshoes commented:
try --no-cache, man
EDIT: ohhh i see, you tried both

teleshoes commented:
i had a typo that called some of the caching stuff even with --no-cache. try again. also, we could discuss this there since this pr is unrelated, really

sillybear commented:
I'm trying your dwi2sm/ssc source patch for Windows but couldn't get it work because it crashes before startup.

If I try --convert D:\Desktop\Songs\Test\BUTTERFLY :
Crash reason: Assertion 'sDirectoryParts.size() >= 4' failed

if I try --convert="D:\Desktop\Songs\Test\BUTTERFLY" :
Crash reason: Access Violation

Also, where should I put the perl script file?

I also, tried to run the perl script alone, changed my converter build to "D:/_old Desktop/SM5 2020/Program", placed the script in StepMania5/Program folder. Seems like it executed but I can't see the output.

teleshoes commented:
note that this pull request is still valid and it would be nice if it got merged until it can be replaced with a more fully-fledged CLI.
all the comments above are regarding scripts outside of the stepmania repo that invoke the stepmania binary.

any other comments/criticisms welcome

lancer1977 reviewed (approved):
Auto-approval by codex operation
